DataFormats/Detectors/FIT/FV0/include/DataFormatsFV0/Digit.h

```diff
@@ -0,0 +1,112 @@
+#define ALICEO2_FV0_DIGIT_H
+
+#include "CommonDataFormat/InteractionRecord.h"
+#include "CommonDataFormat/TimeStamp.h"
+#include <iosfwd>
+#include "Rtypes.h"
+
+namespace o2
+{
+namespace fv0
+{
+
+struct ChannelData {
+  //Int_t ChId;       // channel Id
+  //Double_t CFDTime; // time in ns, 0 at lhc clk center
+  //Double_t QTCAmpl; // Amplitude in mips
+  //int numberOfParticles;
+  Int_t mPmtNumber;   // PhotoMultiplier number (0 to 47)
+  Float_t mTime;      // Time of Flight
+  Short_t mChargeAdc; // ADC sample as present in raw data
+  //Bit information from FEE
+  //Bool_t mIntegrator;
+  //Bool_t mDoubleEvent;
+  //Bool_t mEvent1TimeLost;
+  //Bool_t mEvent2TimeLost;
+  //Bool_t mAdcInGate;
+  //Bool_t mTimeTooLate;
+  //Bool_t mAmpTooHigh;
+  //Bool_t mEventInTrigger;
+  //Bool_t mTimeLost;
+
+  ClassDefNV(ChannelData, 1);
+};
+
+/// \class Digit
+/// \brief FIT digit implementation
+using DigitBase = o2::dataformats::TimeStamp<double>;
+class Digit : public DigitBase
+{
+ public:
+  Digit() = default;
+
+  Digit(std::vector<ChannelData> ChDgDataArr, Double_t time, uint16_t bc, uint32_t orbit, std::vector<Bool_t> const& triggers)
+  {
+    setChDgData(std::move(ChDgDataArr));
+    setTime(time);
+    setInteractionRecord(bc, orbit);
+    setTriggers(std::move(triggers));
+  }
+
+  ~Digit() = default;
+
+  Double_t getTime() const { return mTime; }
+  void setTime(Double_t time) { mTime = time; }
+
+  void setInteractionRecord(uint16_t bc, uint32_t orbit)
+  {
+    mIntRecord.bc = bc;
+    mIntRecord.orbit = orbit;
+  }
+
+  const o2::InteractionRecord& getInteractionRecord() const { return mIntRecord; }
+  o2::InteractionRecord& getInteractionRecord(o2::InteractionRecord& src) { return mIntRecord; }
+  void setInteractionRecord(const o2::InteractionRecord& src) { mIntRecord = src; }
+  uint32_t getOrbit() const { return mIntRecord.orbit; }
+  uint16_t getBC() const { return mIntRecord.bc; }
+
+  const std::vector<Bool_t>& GetTriggers() const { return mTriggers; }
+  std::vector<Bool_t>& GetTriggers() { return mTriggers; }
+  void setTriggers(const std::vector<Bool_t>& triggers) { mTriggers = triggers; }
+  void setTriggers(std::vector<Bool_t>&& triggers) { mTriggers = std::move(triggers); }
+
+  const std::vector<ChannelData>& getChDgData() const { return mChDgDataArr; }
+  std::vector<ChannelData>& getChDgData() { return mChDgDataArr; }
+  void setChDgData(const std::vector<ChannelData>& ChDgDataArr) { mChDgDataArr = ChDgDataArr; }
+  void setChDgData(std::vector<ChannelData>&& ChDgDataArr) { mChDgDataArr = std::move(ChDgDataArr); }
+
+  void printStream(std::ostream& stream) const;
+
+  void cleardigits()
+  {
+    mTriggers.clear();
+    mChDgDataArr.clear();
+  }
+
+ private:
+  Double_t mTime;                   // time stamp
+  o2::InteractionRecord mIntRecord; // Interaction record (orbit, bc)
+
+  //online triggers processed on TCM
+
+  std::vector<Bool_t> mTriggers; // TODO: unsued until triggers of FV0 are defined
```

Comment:
Using vectors of vectors is very inefficient from the memory management and DPL messaging points of view. Could you consider changing the digit class from a vector of channelData to a global vector of channelData (for the whole time frame) and a separate vector of structs with the int.record and reference on the range in the vector<channelData> ? 
As an example you can use [DataFormatsZDC/ChannelData.h as an equivalent of your ChannelData](https://github.com/AliceO2Group/AliceO2/blob/dev/DataFormats/Detectors/ZDC/include/DataFormatsZDC/ChannelData.h) and the [DataFormatsZDC/BCData](https://github.com/AliceO2Group/AliceO2/blob/dev/DataFormats/Detectors/ZDC/include/DataFormatsZDC/BCData.h) as a wrapper for the digit.
I was going to ask also @AllaMaevskaya to introduce similar changes in the T0.

Reply:
Hi @shahor02 ,
I have a few questions about the `o2::zdc::BCData` members.
1. Is the variable `uint32_t channels = 0` used as a flag system? In other words do you store only the values for channels which were non-zero and this variable is treated as a 32-bit flag system to indicate which channels were non-zero and stored in the digits file?
2. The follow-up question would be: if in the beginning I would like to skip such optimization and always save all 48 channels of FV0, I could completely skip this variable, right?
3. Then I had a glimpse into other ZDC digitizer-related classes. You use there cached information from several BCs from the past. Is this only for triggering, or is it needed for something else?

Reply:
Hi @mslupeck 
1) the ``channels`` I added just as a bit pattern of the channels (0:25) stored for this BC in the accompanying vector<ChannelData> (similarly, triggers pattern gives those channels which triggered this BC). It does not affect the access to data, just allows a fast query w/o actually inspecting the content of vector<ChannelData>.
2) right, but for the fixed size class you actually don't need such a parallel containers.
3) the caching is done to account for eventual pile-up: single collision may affect a few BCs which may get extra contribution from the following collisions. In case of ZDC this is further complicated by the requirement to read 2 BCs preceding the (self-)trigger (4 BCs in externally-triggered mode). The T0 digitization also does the caching, is not it more appropriate for V0?

Cheers,
 Ruben

Reply:
Hi @shahor02 
1. OK, thanks
2. Could you elaborate? Which parallel containers are you talking about? 
3. Yes, pile-up will also affect V0, but I was thinking to postpone its implementation for the next PR (especially that it is not very clear to me yet how it should be handled)

Reply:
Hello
T0 fills Digits for 1 BC (Interaction Record) only. What is looks like cache is filling Digits step-by-step: simulate CFD response -> set charge -> set  triggers -> write result. 

Best regards
Alla

Reply:
@AllaMaevskaya I thought the T0 digitization already supports the pile-up. So far so good... On the other front, is there any progress in getting rid of vector of vectors?

@mslupeck for the (2): ZDC reads out 26 channels (``o2::zdc::ChannelData``) per BC, since only few of them are fired, to save the space we store only them. The channels information is placed into the ``vector<o2::zdc::ChannelData>`` (one vector for the whole TF). The belonging of the channels to particular BC is stored in the accompanying vector<o2::zdc::BCData>, the ``BCData`` contains (apart from the orbit/bc and ZDC specific bitfields) an index of its 1st ChannelData entry in the ``vector<o2::zdc::ChannelData>`` and the number of such entries.
These entries can be retrieved from the BCData object via method https://github.com/AliceO2Group/AliceO2/blob/569057e0f47e6708bfdc459d9640613bcd6a1556/DataFormats/Detectors/ZDC/include/DataFormatsZDC/BCData.h#L47
See [macro](https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/ZDC/macro/readZDCDigits.C) for the way the data are accessed.
The same mechanism is implemented for the ITS digits, clusters, tracks etc.: there is a "readout frame record [ROFRecord](https://github.com/AliceO2Group/AliceO2/blob/dev/DataFormats/Detectors/ITSMFT/common/include/DataFormatsITSMFT/ROFRecord.h) class which points on the 1st entry and the number of entries in the ``vector<o2::itsmft::Digit>`` etc. classes.
This mechanism allows to use POD objects only (good for shared memory use and no-copy data transfer).

Reply:
Hello
T0 electronics registered only 1 event per bunch, does not matter where it comes from. PM data has flag - 2 time hits, but in this case amplitude is not correct.

Reply:
Hi @AllaMaevskaya sorry, did not understand what I meant here. To me there are 2 problems to solve (both for V0 and T0): 
1) get rid of the vector in every digit, since this leads to a vector of vectors on the TF container level. This is easy to achieve with the mechanism I used for the ZDC or ITS. Note that internally (during the digitization), you are free to use vectors or other dynamical memory containers for the accumulation of contributions to single digits (obviously, deleting/creating/expanding such containers will have its cost).
2) Digitization of the pile-up: account for the tracks from different collisions contributing to the same BC readout (unique digit per PB in your terms), either from the interaction pile-up in the same BC or due to the slow particles from prior interactions (not sure it applies to T0 but certainly does to V0).

Reply:
OK , do I understand now:
1. you want from me to write 1 gsl span of Channel data for TF and just keep number of 1st entry ( and N channels) for each bunch (InteractionRecord) ?  
2. we do not write tracks. We are storing in MClabels number of charge particles with its source per readout channel. If particle hits radiator in the middle of the bunch +-2ns - we will register it. It can come from any interaction within one bunch or be slow one from  prior interactions - we have no possibility to check this. If there were more than 1 time hit in 1 channel within bunch PM (processing module) will set flag in raw data.

Reply:
@AllaMaevskaya : 
1) almost: in the 1st vector (which then can be provider to other DPL devices as a span) you write all the channels of the TF. In the 2nd vector you write for every trigger of BC (your digit) the index of the 1st channel of this trigger in the 1st vector and the number of (contiguous) channels. Also, the entry of this 2nd vector contains the general info like the IR etc.
2) Exactly: now single MC event contributes to just 1 digit, if this event had slow particles hitting detector during the next BC, these hits will be lost, while with the correct pile-up implementation they should appear in the "digit" of the next event.

Reply:
1. Ok, now it more clear

2. But source ID will be the same as for good hits, yes? We can estimate contribution of slow particles, I don't think real effect due to very narrow registration time range.

Reply:
@AllaMaevskaya yes, the sourceID is the same, but the eventID is different. Losing the slow particles might have negligible effect for the T0, but not the same-BC pile-up, which might be quite high in pp.

Reply:
@shahor02 Thanks for all the explanations - they made it more clear. I have hopefully done what you requested in point 1, ie. there is no vector of vectors anymore. However the pileup (taking into account slow particles) is still not impemented correctly. Triggers are also missing.

I would suggest to postpone implementation of the pileup and triggers, as it is not clear yet how FV0 electronics will actually generate triggers or what the acceptance/charge integration windows will be.